src/Scrutor/ServiceCollectionExtensions.Decoration.cs

```diff
@@ -343,5 +348,51 @@ private static object CreateInstance(this IServiceProvider provider, Type type,
         {
             return ActivatorUtilities.CreateInstance(provider, type, arguments);
         }
+
+        private static Type GetImplementationType(this ServiceDescriptor serviceDescriptor)
+        {
+            if (serviceDescriptor.ImplementationType != null)
+                return serviceDescriptor.ImplementationType;
+
+            if (serviceDescriptor.ImplementationInstance != null)
+                return serviceDescriptor.ImplementationInstance.GetType();
+
+            if (serviceDescriptor.ImplementationFactory != null)
+                return serviceDescriptor.ImplementationFactory.GetType().GenericTypeArguments[1];
```

Comment:
The generic type arg is always `Object` because the signature of the factory is `Func<IServiceProvider,object>`.

```
public interface IFoo { }
public class Foo : IFoo { }

var descriptor = ServiceDescriptor.Transient(typeof(IFoo), prov => new Foo());
var type = descriptor.GetImplementationType(); // return Object
```

Having a factory we cannot determine the implementation type because there can be multiple and some of them might be disposable and some are not:

```
public class Foo2 : IFoo, IDisposable
{
	public void Dispose()	{ 	}
}

var descriptor1 = ServiceDescriptor.Transient(typeof(IFoo), prov =>
{
	var foo = new Random().Next() == 42
				       ? (IFoo) new Foo()
  				       : new Foo2();
		
	return foo;
});
```

Reply:
>The generic type arg is always `Object` because the signature of the factory is `Func<IServiceProvider,object>`.

Actually, that;s not the case, as `GetType()` gets the runtime type. However it still doesn't work for other reasons 🙂 In the example you've provided, the factory is of type `Func<IServiceProvider, IFoo>`, so `descriptor.GetImplementationType()` returns `IFoo`, not `object` as you suggest, but also not `Foo` as intended!

> Having a factory we cannot determine the implementation type because there can be multiple and some of them might be disposable and some are not.

Hmmm, yeah that's a very good point, I think the specific approach in this PR falls down fundamentally because of that.